dlib/image_transforms/interpolation.h

```diff
@@ -965,6 +965,46 @@ namespace dlib
         swap(img, temp);
     }
 
+// ----------------------------------------------------------------------------------------
+
+    template <typename image_type>
+    point_transform_affine letterbox_image (
+        const image_type& img_in,
+        image_type& img_out,
+        long size
+    )
+    {
+        DLIB_CASSERT(size > 0, "size must be bigger than zero, but was " << size);
+        // the scaling factor of the image
+        const auto scale = size / std::max<double>(img_in.nr(), img_in.nc());
+
+        // early return if the image has already the requested size and no padding is needed
+        if (scale == 1 and img_in.nr() == img_in.nc())
+        {
+            img_out = img_in;
```

Comment:
This won't work for `array2d` or other non-copyable images.  You want to use `assign_image()`

Reply:
Yeah I guess the code is assuming the image is a matrix.  Make this use the generic image interface and it's all good.  Like use `const_image_view` and `image_view` to access the image contents.